### ISSUE TITLE
Building pre-processor against mdBook 0.4.6 fails

### ISSUE THREAD

**Issue by plul** [rust-lang/mdBook#1448]
It works against 0.4.5 and prior versions. Against 0.4.6 I get this:

```
error[E0407]: method `supports_renderer` is not a member of trait `Preprocessor`
   --> /home/<user>/.cargo/registry/src/github.com-1ecc6299db9ec823/mdbook-0.4.6/src/preprocess/cmd.rs:126:5
    |
126 | /     fn supports_renderer(&self, renderer: &str) -> bool {
127 | |         debug!(
128 | |             "Checking if the \"{}\" preprocessor supports \"{}\"",
129 | |             self.name(),
...   |
164 | |         outcome.unwrap_or(false)
165 | |     }
    | |_____^ not a member of trait `Preprocessor`

error[E0599]: no method named `supports_renderer` found for reference `&dyn Preprocessor` in the current scope
   --> /home/<user>/.cargo/registry/src/github.com-1ecc6299db9ec823/mdbook-0.4.6/src/book/mod.rs:446:29
    |
446 |         return preprocessor.supports_renderer(renderer.name());
    |                             ^^^^^^^^^^^^^^^^^ method not found in `&dyn Preprocessor`

error[E0599]: no method named `supports_renderer` found for reference `&dyn Preprocessor` in the current scope
   --> /home/<user>/.cargo/registry/src/github.com-1ecc6299db9ec823/mdbook-0.4.6/src/book/mod.rs:459:18
    |
459 |     preprocessor.supports_renderer(renderer_name)
    |                  ^^^^^^^^^^^^^^^^^ method not found in `&dyn Preprocessor`
```

The pre-processor is unfortunately not open source, so I am sorry that I cannot share that with you. I hope the above is enough to help you.

**Comment by ehuss** [comment]
Hm, that's really strange!  I can't see anything that would explain that. Looking at [the diff](https://github.com/rust-lang/mdBook/compare/v0.4.5...v0.4.6), I don't see anything that touches any public API or anything that touches the Preprocessor trait at all.  I also created a new project that implements the [`nop-preprocessor`](https://github.com/rust-lang/mdBook/blob/master/examples/nop-preprocessor.rs), and it built fine, as well as some other preprocessors.

Can you maybe put together a small reproduction that you can share?  Also, which version of rustc are you using?

**Comment by plul** [comment]
Thank you for looking into it at quickly!

I will try to put together a small example if I can, or git bisect to find out which commit causes it. 
I probably won't be able to do so until Monday or Tuesday, however.

**Comment by plul** [comment]
```
❯ rustc --version
rustc 1.48.0 (7eac88abb 2020-11-16)
```

**Comment by plul** [comment]
This morning the compilation is working. I Ultimately cannot tell what the problem was, I am sorry.
In case you are interested, the following is a description of what I did and found out along the way.

To debug, I started stripping things out of my preprocessor, until I could no longer tell the difference between it and the `nop-preprocessor`. Still building against `0.4.6` produced the error above. Then, I did the following

* Updated `rustc` to `1.49.0`
* Ran `cargo clean`
* Removed `Cargo.lock`

I also nuked my `~/.cargo` directory at some point for good measure.

After that I was getting a new kind of error! Similar looking errors as [these](https://github.com/serde-rs/serde/issues/1950) [folks](https://github.com/serde-rs/serde/issues/1949), related to `serde`. I suspected that mdBook `0.4.6` depended on private things in `serde` and `serde` `1.0.119` broke it, but I do not think that is true, because I was just about to report these findings to you, but I went to double check: I ran `cargo clean` and `rm Cargo.lock` again and suddenly the build was working!

I am sorry I do not know what is up. It now works with mdBook `0.4.5` and `0.4.6` and I can no longer reproduce the error. My best guess is that mdbook depends on `something` that depended on `serde` and broke when `serde` updated to `1.0.119` and while I was debugging, they released a fix for `something`, so my `cargo clean; rm Cargo.lock`, suddenly made things work.
